### ISSUE TITLE
linter: react/immutability false positives on react-native-reanimated shared value writes

### ISSUE THREAD

**Issue by shiroyasha9** [oxc-project/oxc#26222]
### What version of Oxlint are you using?

1.80.0

### What command did you run?

oxlint@1.80.0 --react-plugin -A all -D react/immutability Fixture.tsx

### What does your `.oxlintrc.json` (or `oxlint.config.ts`) config file look like?

No config file. The reproduction uses CLI flags only, so the report does not depend on our repo's config.

### What happened?

We're evaluating oxlint for our Expo/React Native monorepo against our current setup, `eslint-plugin-react-hooks` running as a React Compiler linter. One thing blocks the switch: every Reanimated shared-value write in the repo reports a false positive.

## Reproduction

`Fixture.tsx`:

```tsx
import {useSharedValue} from 'react-native-reanimated';

function SomeComponent() {
  const sharedVal = useSharedValue(0);
  return (
    <Button
      onPress={() => (sharedVal.value = Math.random())}
      title="Randomize"
    />
  );
}
```

Error:

```
Fixture.tsx:7:23: error react(immutability): This value cannot be modified
  help: Modifying a value returned from a hook is not allowed. Consider moving the
        modification into the hook where the value is constructed
Fixture.tsx:7:23: error react(immutability): Cannot modify local variables after render completes
  help: This argument is a function which may reassign or mutate `sharedVal` after render,
        which can cause inconsistent behavior on subsequent renders. Consider using state instead
```

Expected: no diagnostics. `eslint-plugin-react-hooks@7.1.1` with `recommended-latest` reports nothing for this file.

The above snippet is React's own compiler fixture, which oxc already vendors as `crates/oxc_react_compiler/fixtures/reanimated-shared-value-writes.jsx`. Oxlint doesn't honour the fixture's `// @enableCustomTypeDefinitionForReanimated` pragma either, so restoring that line changes nothing and there's no in-file workaround.

## Why the code is valid

A Reanimated shared value is a mutable container that lives outside React state. Writing to `sharedVal.value` outside render is the documented API, and it's how every `useSharedValue` animation is driven. React gives the module a custom type model so the compiler treats those values as ref-like.

## How babel-plugin-react-compiler handles it

The compiler ships that model and switches it on by detection rather than configuration. `Entrypoint/Reanimated.ts` exports `pipelineUsesReanimatedPlugin()`, which scans the Babel plugin list for a `react-native-reanimated` key and then falls back to `require.resolve('react-native-reanimated')`. If either check hits, `injectReanimatedFlag()` sets `environment.enableCustomTypeDefinitionForReanimated: true`. All of this runs in `BabelPluginReactCompiler`'s `Program.enter`, behind `enableReanimatedCheck`, which defaults to `true`.

`eslint-plugin-react-hooks` gets the same behaviour for free, because its lint path runs the compiler through that same Babel plugin (`plugins: [[BabelPluginReactCompiler, options]]` in `runReactCompilerImpl`). So an ESLint user working in a Reanimated project never configures anything, which is why the fixture above comes back clean.

## What's missing in oxlint

The model itself is already ported, and you have tests for it:

- `oxc_react_compiler::react_compiler_hir::globals::get_reanimated_module_type`
- `REANIMATED_SHARED_VALUE_ID` in `object_shape.rs`, which `is_ref_or_ref_like_mutable_type` honours
- `environment.rs`, which registers the `react-native-reanimated` module type when the flag is set
- fixtures and snapshots for both the enabled and the non-imported cases

The flag never becomes `true` once the linter drives the compiler. `EnvironmentConfig::default()` sets `enable_custom_type_definition_for_reanimated: false`, and `crates/oxc_linter/src/utils/react_compiler.rs` builds a fixed `EnvironmentConfig` for the whole rule family without ever setting that field. Neither of Babel's two signals is available to oxlint anyway: there's no Babel plugin list to inspect, and oxlint only builds a resolver when `cross_module` is set, so the default lint path has nothing to resolve with.

## Impact

On our monorepo (about 2,000 files), running the Oxlint equivalents of the official recommended compiler categories:

| | diagnostics | `react/immutability` |
| --- | ---: | ---: |
| oxlint 1.80.0 | 32 | 20 |
| `eslint-plugin-react-hooks` 7.1.1, same categories | 0 | 0 |

18 of those 20 `immutability` reports are Reanimated `.value` writes, spread across animation hooks, drawers, skeletons and scroll handlers.

## Possible directions

The type model is keyed on the import, so switching it on does nothing to code that never imports Reanimated. Your `error.invalid-non-imported-reanimated-shared-value-writes.jsx` fixture shows this, since it sets the flag, calls a global `useSharedValue`, and still reports `This value cannot be modified`.

**1. Switch it on in the linter's compiler options.**

- Changes one default, with no configuration and no new file access.
- Projects that never import Reanimated are unaffected, per the fixture above.
- There's precedent for the mechanism: `react_compiler_plugin_options()` already overrides `EnvironmentConfig::default()` in ten places, including `validate_no_set_state_in_effects`, `validate_static_components` and `validate_no_void_use_memo`. Setting this flag there would be one more field in that struct literal.
- One caveat: `enableCustomTypeDefinitionForReanimated` isn't part of upstream's `COMPILER_OPTIONS`, because upstream sets it by detection instead. 

**2. Detect it per file,** enabling the model for files that import `react-native-reanimated`.

- This keys on the same thing the type model already keys on.
- The `Semantic` is already available where the compiler runs.
- There's precedent: 23 rule files branch on `ctx.module_record().import_entries`, and `react/jsx_key.rs` has an `import_matcher()` helper for it. `should_run_react_compiler()` also gates the family per file already, on path and extension.

**3. Resolve the package,** mirroring Babel's `require.resolve` fallback so the behaviour matches exactly.

- Oxlint does embed `oxc_resolver`, but `Runtime::get_resolver()` only builds a resolver when `cross_module` is set.
- The model would then depend on `--import-plugin` without saying so, which feels like the wrong thing to couple a React Native default to.

**4. Make it explicit,** as a `react` plugin setting or a rule option.

- This leaves the default broken for every Reanimated project until someone finds the option, so I think it works better as an override than as the fix.
- A rule option would also break the invariant that one shared compiler run stays valid for any combination of enabled rules, since it changes what the compiler analyzes rather than which categories get reported.
- There's precedent: `settings.react.version`, `settings.react.componentWrapperFunctions` and `settings.jsx-a11y.components` are all read this way.

Options 1 to 3 restore parity with `babel-plugin-react-compiler`, where a React Native developer never has to learn that this flag exists.

I'm happy to implement if you point me at a direction. I've already prototyped option 4 locally, threading the flag through `ReactPluginSettings` into the shared `EnvironmentConfig`, and it takes our 32 diagnostics down to 14 with nothing else changed. That was mostly to check how small the plumbing is, though. 

## Other versions

- `eslint-plugin-react-hooks` 7.1.1 (comparison)
- `react-native-reanimated` 4.5.1
- macOS arm64

---

*Issue improved with AI assistance. I verified every claim above by running both linters and reading the referenced source.*
